File: cpp/problem88.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

/**
 * ordinary merge operation
 * 
 * Here the space of nums1 should be stolen to record the result.
 * So we move the first m elements in nums1 to the back of this array,
 * to avoid that they are overwrite during merger.
 */ 
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        moveBack(nums1, m, n);
        int i = n, j = 0, mpn = m + n, k = 0;
        while( i < mpn && j < n ){
            if( nums1[i] < nums2[j] ) nums1[k++] = nums1[i++];
            else nums1[k++] = nums2[j++];
        }
        if( i == mpn ) while( j < n ) nums1[k++] = nums2[j++];
    }
    void moveBack(vector<int> &nums1, int m, int n ){
        for(int i=m-1; i>=0; --i) nums1[i+n] = nums1[i];
    }
};



int main(int argc, char const *argv[]){
    Solution sol;

    // test case in Leetcode Problem
    vector<int> nums1 = {1,2,3,0,0,0};
    vector<int> nums2 = {2,5,6};

    sol.merge(nums1, 3, nums2, 3);
    MYFILE::print_range(nums1) << endl;

    // case in Leetcode Judgement System.
    nums1.assign({1,2,4,5,6,0});
    nums2.assign({3});
    sol.merge(nums1, 5, nums2, 1);
    MYFILE::print_range(nums1) << endl;

    return 0;
}

File: swift/problem88.swift
class Solution {
    func merge(_ nums1: inout [Int], _ m: Int, _ nums2: [Int], _ n: Int) {
        var tail = m+n-1; var p1 = m-1; var p2 = n-1
        
        while (p1 >= 0 || p2 >= 0) {
            if (p1 >= 0 && p2 >= 0) {
                if (nums1[p1] >= nums2[p2]) {
                    nums1[tail] = nums1[p1]
                    p1 -= 1
                } else {
                    nums1[tail] = nums2[p2]
                    p2 -= 1
                }
                tail -= 1
            } else if (p2 >= 0) {
                while (p2 >= 0) {
                    nums1[tail] = nums2[p2]
                    p2 -= 1; tail -= 1
                }
                return
            } else {
                return
            }
            
        }
    }
}
